File: GymApp/composeApp/src/commonMain/kotlin/com/gym/gymapp/data/repository/PlanRepository.kt
package com.gym.gymapp.data.repository

import com.gym.gymapp.data.models.MembershipPlan
import com.gym.gymapp.network.NetworkClient
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.serialization.Serializable

import com.gym.gymapp.data.models.ApiResponse
import com.gym.gymapp.ui.components.NotificationManager
import com.gym.gymapp.ui.components.AppNotificationType

class PlanRepository {
    private val client = NetworkClient.client

    suspend fun getPlans(gymId: String): List<MembershipPlan> {
        val response = client.get("/plans") {
            parameter("gymId", gymId)
        }
        return if (response.status == HttpStatusCode.OK) {
            val apiResponse: ApiResponse<List<MembershipPlan>> = response.body()
            if (!apiResponse.status) {
                NotificationManager.showNotification(apiResponse.message, AppNotificationType.ERROR)
                emptyList()
            } else {
                apiResponse.data
            }
        } else {
            emptyList()
        }
    }


    suspend fun createPlan(plan: MembershipPlan): Boolean {
        val response = client.post("/plans") {
            contentType(ContentType.Application.Json)
            setBody(plan)
        }
        if (response.status == HttpStatusCode.Created || response.status == HttpStatusCode.OK) {
            val apiResponse: ApiResponse<MembershipPlan> = response.body()
            if (!apiResponse.status) {
                NotificationManager.showNotification(apiResponse.message, AppNotificationType.ERROR)
            }
            return apiResponse.status
        }
        return false
    }

    suspend fun updatePlan(planId: String, plan: MembershipPlan): Boolean {
        val response = client.patch("/plans/$planId") {
            contentType(ContentType.Application.Json)
            setBody(plan)
        }
        return response.status == HttpStatusCode.OK
    }

    suspend fun deletePlan(planId: String): Boolean {
        if (planId.isBlank()) return false
        val response = client.delete("/plans/$planId")
        return response.status == HttpStatusCode.OK || response.status == HttpStatusCode.NoContent
    }
}

File: gym_flutter_app/lib/data/repositories/plan_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_client.dart';
import '../models/api_response.dart';

class MembershipPlan {
  final String id;
  final String name;
  final double price;
  final int durationMonths;
  final String? description;
  final String gymId;

  MembershipPlan({
    required this.id,
    required this.name,
    required this.price,
    required this.durationMonths,
    this.description,
    required this.gymId,
  });

  factory MembershipPlan.fromJson(Map<String, dynamic> json) {
    return MembershipPlan(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      durationMonths: json['durationMonths'] ?? 0,
      description: json['description'],
      gymId: json['gymId'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'price': price,
    'durationMonths': durationMonths,
    'description': description,
    'gymId': gymId,
  };
}

class PlanRepository {
  final Dio _client;

  PlanRepository(this._client);

  Future<ApiResponse<List<MembershipPlan>>> getPlans(String gymId) async {
    try {
      final response = await _client.get('/plans', queryParameters: {'gymId': gymId});
      return ApiResponse.fromJson(
        response.data,
        (data) => (data as List).map((e) => MembershipPlan.fromJson(e)).toList(),
      );
    } catch (e) {
      return ApiResponse(status: false, message: e.toString(), data: []);
    }
  }

  Future<ApiResponse<MembershipPlan>> createPlan(MembershipPlan plan) async {
    try {
      final response = await _client.post('/plans', data: plan.toJson());
      return ApiResponse.fromJson(
        response.data,
        (data) => MembershipPlan.fromJson(data),
      );
    } catch (e) {
      return ApiResponse(status: false, message: e.toString(), data: null);
    }
  }

  Future<ApiResponse<bool>> deletePlan(String planId) async {
    try {
      final response = await _client.delete('/plans/$planId');
      return ApiResponse.fromJson(response.data, (data) => true);
    } catch (e) {
      return ApiResponse(status: false, message: e.toString(), data: false);
    }
  }
}

final planRepositoryProvider = Provider((ref) => PlanRepository(ref.watch(apiClientProvider)));
